src/common/actions/repositories.js

```diff
@@ -53,8 +54,124 @@ export function fetchUserRepositories(pageNumber) {
 
 export function fetchRepositoriesError(error) {
   return {
-    type: FETCH_REPOSITORIES_ERROR,
+    type: REPOSITORIES_FAILURE,
     payload: error,
     error: true
   };
 }
+
+// XXX may need to split out the fetch from the success and failure dispatch
+// to ensure than setRepositories happens in sync with refreshed user snaps
+export function fetchUserRepositoriesAndSnaps(owner) {
+  return (dispatch) => {
+    return Promise.all([
+      dispatch(fetchUserRepositories()),
+      dispatch(fetchUserSnaps(owner))
+    ]);
+  };
+}
+
+// XXX move to repository actions?
+// could be plural or singular, could be in one module or two, might be clearer in one for now
+export const REPOSITORY_TOGGLE_SELECT = 'REPOSITORY_TOGGLE_SELECT';
+
+export function toggleRepositorySelection(id) {
+  return {
+    type: REPOSITORY_TOGGLE_SELECT,
+    payload: {
+      id
+    }
+  };
+}
+
+export const REPOSITORY_BUILD = 'REPOSITORY_BUILD';
+export const REPOSITORY_SUCCESS = 'REPOSITORY_SUCCESS';
+export const REPOSITORY_FAILURE = 'REPOSITORY_FAILURE';
+export const REPOSITORY_RESET = 'REPOSITORY_RESET';
+
+export function buildRepositories(repositories) {
+  return (dispatch) => {
+    const promises = repositories.map(
+      (repository) => {
+        return dispatch(buildRepository(repository));
+      }
+    );
+    return Promise.all(promises).then(() => dispatch(push('/dashboard')));
+  };
+}
+
+export function buildRepository(repository) {
+  const { id, url, owner, name } = repository;
+
+  return async (dispatch) => {
+    dispatch({
+      type: REPOSITORY_BUILD,
+      payload: {
+        id
+      }
+    });
+
+    try {
+      await createWebhook(owner, name);
+      const response = await fetch(`${BASE_URL}/api/launchpad/snaps`, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ repository_url: url }),
+        credentials: 'same-origin'
+      });
+
+      await checkStatus(response);
+      dispatch(buildRepositorySuccess(id));
+    } catch (error) {
+      dispatch(buildRepositoryError(id, error));
+      return Promise.reject(error);
```

Comment:
Good question. The pattern we have is that if an async action succeeds we do `X`, and if it fails we do `Y`. Where we have batch actions we switch to then looking at state, which is valid but has some complexities, but why not just continue the same pattern? If the batch action is successful we do 'Z', etc.

The complexity of looking at state for errors is, I think, that you need to then include `success` in state, or some indication that you performed an action on the item, because checking for errors doesn't tell you if an action was ever performed (a fresh list has no errors), so in the case where you want to redirect if everything was successful, it just seems easier to stick to the the resolve/reject pattern than to decorate state.